File: care/emr/reports/context_builder/data_points/payment_reconciliation.py
from django_filters import rest_framework as filters

from care.emr.models.payment_reconciliation import PaymentReconciliation
from care.emr.reports.context_builder.data_points.base import (
    Field,
    QuerysetContextBuilder,
)
from care.emr.reports.context_builder.data_points.invoice import (
    MinimumInvoiceContextBuilder,
)

PAYMENT_RECONCILIATION_STATUS = {
    "active": "Active",
    "cancelled": "Cancelled",
    "draft": "Draft",
    "entered_in_error": "Entered in Error",
}

PAYMENT_RECONCILIATION_TYPE = {
    "payment": "Payment",
    "adjustment": "Adjustment",
    "advance": "Advance",
}
PAYMENT_RECONCILIATION_KIND = {
    "deposit": "Deposit",
    "periodic_payment": "Periodic Payment",
    "online": "Online",
    "kiosk": "Kiosk",
}
PAYMENT_RECONCILIATION_ISSUER_TYPE = {
    "patient": "Patient",
    "insurer": "Insurer",
}
PAYMENT_RECONCILIATION_OUTCOME = {
    "queued": "Queued",
    "complete": "Complete",
    "error": "Error",
    "partial": "Partial",
}
PAYMENT_RECONCILIATION_PAYMENT_METHOD = {
    "cash": "Cash",
    "ccca": "Credit Card",
    "cchk": "Credit Check",
    "cdac": "Credit Account",
    "chck": "Check",
    "ddpo": "Direct Deposit",
    "debc": "Debit Card",
}


class PaymentReconciliationReportFilter(filters.FilterSet):
    status = filters.CharFilter(lookup_expr="iexact")
    target_invoice = filters.UUIDFilter(field_name="target_invoice__external_id")
    reconciliation_type = filters.CharFilter(lookup_expr="iexact")
    is_credit_note = filters.BooleanFilter(field_name="is_credit_note")
    location = filters.UUIDFilter(field_name="location__external_id")


class PaymentReconciliationContextBuilder(QuerysetContextBuilder):
    filterset_class = PaymentReconciliationReportFilter
    __filterset_backends__ = [filters.DjangoFilterBackend]

    status = Field(
        display="Payment Reconciliation Status",
        preview_value="Active",
        mapping=lambda p: PAYMENT_RECONCILIATION_STATUS.get(p.status, p.status.title())
        if p.status
        else "",
        description="Status of the payment reconciliation",
    )
    reconciliation_type = Field(
        display="Reconciliation Type",
        preview_value="Payment",
        mapping=lambda p: PAYMENT_RECONCILIATION_TYPE.get(
            p.reconciliation_type, p.reconciliation_type.title()
        )
        if p.reconciliation_type
        else "",
        description="Type of the payment reconciliation",
    )
    amount = Field(
        display="Reconciled Amount",
        preview_value="150.00",
        description="Amount reconciled in the payment reconciliation",
    )
    reference_number = Field(
        display="Reference Number",
        preview_value="PR-1001",
        description="Unique reference number of the payment reconciliation",
    )
    kind = Field(
        display="Kind",
        preview_value="Credit",
        mapping=lambda p: PAYMENT_RECONCILIATION_KIND.get(p.kind, p.kind.title())
        if p.kind
        else "",
        description="Kind of payment reconciliation",
    )

    is_credit_note = Field(
        display="Is Credit Note",
        preview_value="False",
        description="Indicates if the reconciliation is a credit note",
    )
    issuer_type = Field(
        display="Issuer Type",
        preview_value="Patient",
        mapping=lambda p: PAYMENT_RECONCILIATION_ISSUER_TYPE.get(
            p.issuer_type, p.issuer_type.title()
        )
        if p.issuer_type
        else "",
        description="Type of the issuer for the payment reconciliation",
    )
    outcome = Field(
        display="Outcome",
        preview_value="Success",
        mapping=lambda p: PAYMENT_RECONCILIATION_OUTCOME.get(
            p.outcome, p.outcome.title()
        )
        if p.outcome
        else "",
        description="Outcome of the payment reconciliation process",
    )
    method = Field(
        display="Payment Method",
        preview_value="Credit Card",
        mapping=lambda p: PAYMENT_RECONCILIATION_PAYMENT_METHOD.get(
            p.method, p.method.title()
        )
        if p.method
        else "",
        description="Method used for the payment reconciliation",
    )
    target_invoice = Field(
        display="Target Invoice",
        target_context=MinimumInvoiceContextBuilder,
        preview_value="",
        description="Invoice associated with the payment reconciliation",
    )
    created_date = Field(
        display="Created Date",
        preview_value="2024-01-01T10:00:00Z",
        description="Timestamp when the payment reconciliation was created",
    )

    def get_context(self):
        return PaymentReconciliation.objects.filter(account=self.parent_context)

File: frontend/src/types/billing/paymentReconciliation/paymentReconciliation.ts
import { AccountRead } from "@/types/billing/account/Account";
import { InvoiceRead } from "@/types/billing/invoice/invoice";
import { LocationRead } from "@/types/location/location";
import { UserReadMinimal } from "@/types/user/user";

export enum PaymentReconciliationType {
  payment = "payment",
  adjustment = "adjustment",
  advance = "advance",
}

export enum PaymentReconciliationStatus {
  active = "active",
  cancelled = "cancelled",
  draft = "draft",
  entered_in_error = "entered_in_error",
}

export const PAYMENT_RECONCILIATION_STATUS_COLORS = {
  active: "primary",
  cancelled: "destructive",
  draft: "secondary",
  entered_in_error: "destructive",
} as const satisfies Record<PaymentReconciliationStatus, string>;

export enum PaymentReconciliationKind {
  deposit = "deposit",
  preriodic_payment = "preriodic_payment",
  online = "online",
  kiosk = "kiosk",
}

export enum PaymentReconciliationIssuerType {
  patient = "patient",
  insurer = "insurer",
}

export enum PaymentReconciliationOutcome {
  queued = "queued",
  complete = "complete",
  error = "error",
  partial = "partial",
}

export const PAYMENT_RECONCILIATION_OUTCOME_COLORS = {
  queued: "secondary",
  complete: "primary",
  error: "destructive",
  partial: "outline",
} as const satisfies Record<PaymentReconciliationOutcome, string>;

export enum PaymentReconciliationPaymentMethod {
  cash = "cash",
  ccca = "ccca",
  cchk = "cchk",
  cdac = "cdac",
  chck = "chck",
  ddpo = "ddpo",
  debc = "debc",
}

export const PAYMENT_RECONCILIATION_METHOD_MAP: Record<
  PaymentReconciliationPaymentMethod,
  string
> = {
  cash: "Cash",
  ccca: "Credit Card",
  cchk: "Credit Check",
  cdac: "Credit Account",
  chck: "Check",
  ddpo: "Direct Deposit",
  debc: "Debit Card",
};

export interface PaymentReconciliationBase {
  id: string;
  reconciliation_type: PaymentReconciliationType;
  status: PaymentReconciliationStatus;
  kind: PaymentReconciliationKind;
  issuer_type: PaymentReconciliationIssuerType;
  outcome: PaymentReconciliationOutcome;
  disposition?: string;
  payment_datetime?: string;
  method: PaymentReconciliationPaymentMethod;
  reference_number?: string;
  authorization?: string;
  tendered_amount?: string;
  returned_amount?: string;
  note?: string;
  amount: string;
}

export interface PaymentReconciliationCreate extends Omit<
  PaymentReconciliationBase,
  "id"
> {
  target_invoice?: string;
  account: string;
  is_credit_note?: boolean;
  location?: string;
  extensions?: Record<string, Record<string, unknown>>;
}

export type PaymentReconciliationUpdate = Omit<PaymentReconciliationBase, "id">;

export interface PaymentReconciliationRead extends PaymentReconciliationBase {
  target_invoice: InvoiceRead;
  account: AccountRead;
  is_credit_note: boolean;
  location: LocationRead | null;
  created_by: UserReadMinimal;
  updated_by: UserReadMinimal;
  created_date: string;
  modified_date: string;
  extensions?: Record<string, Record<string, unknown>>;
}

export interface PaymentReconciliationCancel {
  reason: PaymentReconciliationStatus;
}
